cli/listener/src/main.rs

```diff
@@ -570,18 +611,69 @@ fn get_success_response(interval_prices: IntervalPrices) -> ResponseBox {
     .boxed()
 }
 
+/// Get a header with `Expires` key.
+/// If the end date for the query is far in the past we set it to expire in 1
+/// week. However, if the end date is in the same epoch, we set it to expire in
+/// the approximate time the next epoch begins.
+fn get_expires_header(
+    now: chrono::DateTime<chrono::Utc>,
+    interval_prices: &IntervalPrices,
+    current_clock: Option<Clock>,
+    date_request_type: &DateRequestType,
+) -> Header {
+    let expires_date = match date_request_type {
+        // From the past, set expiration to 1 week from now.
+        DateRequestType::Past => now + chrono::Duration::weeks(1),
+        DateRequestType::MovingTarget => {
+            // We don't have information about the current clock, default to 1hr
+            // cache.
+            match current_clock {
+                None => now + chrono::Duration::hours(1),
+                Some(clock) => {
+                    let epoch_sched = EpochSchedule::without_warmup();
+                    let first_slot_next_epoch =
+                        epoch_sched.get_first_slot_in_epoch(clock.epoch + 1);
+                    let slots_diff = first_slot_next_epoch - clock.slot;
+                    if interval_prices.duration_epochs() > 0 {
+                        // We can estimate the epoch duration.
+                        let epoch_duration = interval_prices.duration_wall_time()
+                            / interval_prices.duration_epochs() as i32;
+                        let slot_duration = epoch_duration
+                            / epoch_sched.get_slots_in_epoch(interval_prices.end_epoch) as i32;
+                        now + slot_duration * (slots_diff * 8 / 10) as i32
```

Comment:
```suggestion
                        // Cache until we estimate that 80% of the current epoch will be complete,
                        // so that we don’t cache for too long if the epoch suddenly ends faster than expected.
                        now + slot_duration * (slots_diff * 8 / 10) as i32
```